### PR TITLE
CI for Linux, MinGW and macOS

### PR DESCRIPTION
**Pull request by illwieckz**
CI for Linux, MinGW and macOS

### CODE REVIEW

**Comment by illwieckz** [comment]
The PR includes #46 to fix the macOS build:

https://github.com/DaemonEngine/crunch/pull/46

**Review by DolceTriade — APPROVED** [review]

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  displayName: 'Setup'
+- bash: |
+    set -xe
+    export CMAKE_BUILD_PARALLEL_LEVEL="$(${NPROC_TOOL})"
+    echo "${CMAKE_BUILD_PARALLEL_LEVEL}"
+    cmake -S. -Bbuild \
+      -G${CMAKE_GENERATOR} \
+      -DCMAKE_C_COMPILER=${C_COMPILER} -DCMAKE_CXX_COMPILER=${CXX_COMPILER} \
+      -DCMAKE_TOOLCHAIN_FILE=${TOOLCHAIN_FILE} \
+      -DBUILD_CRUNCH=ON -DBUILD_SHARED_LIBCRN=ON -DBUILD_EXAMPLES=OFF
```
[the remark this answers is not in the document]
Comment (illwieckz) [reply azure-pipelines.yml:83]: They only builds on Windows as far as I know.

I don't know yet if they build on MinGW.

[the remark this answers is not in the document]
Comment (illwieckz) [reply azure-pipelines.yml]: They will build once #49 is merged:

- https://github.com/DaemonEngine/crunch/pull/49

I now enabled the building of examples in the MinGW jobs, we will have to merge #49 first (which is already required for other MinGW fixes).

File: cmake/cross-toolchain-mingw32.cmake [hunk 2]
```diff
@@ -2,8 +2,7 @@
 set( CMAKE_SYSTEM_NAME Windows )
 set( CMAKE_SYSTEM_PROCESSOR x86 )
 
-# C/C++ compilers
-set( CMAKE_C_COMPILER i686-w64-mingw32-gcc )
```
Comment (slipher) [inline cmake/cross-toolchain-mingw32.cmake:6]: I don't think it helps anything to provide an incomplete version of the toolchain file.

Comment (illwieckz) [reply cmake/cross-toolchain-mingw32.cmake:6]: This tool only use C++, I modified the CMake file to do:

```
project(${CRUNCH_PROJECT_NAME} LANGUAGES CXX)
```

Meaning that even CMake would not look for a C compiler to begin with.

Comment (illwieckz) [reply cmake/cross-toolchain-mingw32.cmake:6]:
```
-- The CXX compiler identification is GNU 12.0.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/x86_64-w64-mingw32-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.40.1") 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Configuring done (0.6s)
-- Generating done (0.0s)
```

Comment (slipher) [reply cmake/cross-toolchain-mingw32.cmake:6]: I think of it like a library component that doesn't need to modified for the application's use case. Like when we have a vendored library, we don't go in and delete all the functions that we don't use.

Comment (illwieckz) [reply cmake/cross-toolchain-mingw32.cmake]: OK, I reverted that line.

**Comment by illwieckz** [comment]
I added some test PNG and TGA test images from:

- https://github.com/UnvanquishedAssets/UnvanquishedTestAssets/tree/master/pkg/map-test-pngalpha_src.dpkdir/textures/test-pngalpha_custom_src
- https://gitlab.com/illwieckz/investigating-tga-orientation-in-imagemagick/-/tree/master/samples

When running the Azure CI on a native platform (no cross-compilation), it runs crunch with them as input.

Currently, the test doesn't verify the output of crunch, but it already checks that the input format is not refused, and would catch this kind of error:

- https://github.com/DaemonEngine/crunch/issues/13

It also makes us able to run the test outside the CI and it makes me discover the TGA input format is not fully supported (horizontal orientation bit is ignored), I guess the error is in STB and then I would have to report them the error.

File: README.md [hunk 3]
```diff
@@ -346,31 +386,31 @@ GPU you're using. (See example2.)
...
 blocks at all, only 4 color blocks. So it doesn't support DXT1A
 transparency, and its output quality suffers a little due to this
 limitation. (Note that the clustered DXTn compressor used when
-writing clustered .DDS files does *not* have this limitation.)
+writing clustered DDS files does _not_ have this limitation.)
 
 * Clustered DXT5/DXT5A compressor is able to group DXT5A blocks into
 clusters only if they use absolute (black/white) selector indices. This
-hurts performance at very low bitrates, because too many bits are
+hurts performance at very low bIt will produce the `crunch.js` and `crunch.wasm` files in the `build-emscripten` directory.itrates, because too many bits are
```
Comment (slipher) [inline README.md:410]: wat

File: README.md [hunk 4]
```diff
@@ -1,17 +1,28 @@
 # Crunch
 
+![Crunch logo](crunch.png)
```
Comment (slipher) [inline README.md:3]: Seems kinda weird for a fork to use a logo belonging to the original version of the project

File: README.md (hunk 3 above)
Comment (illwieckz) [reply README.md:410]: lol yeah, probably a random middle click XD.

File: README.md (hunk 4 above)
Comment (illwieckz) [reply README.md]: I had mixed feeling about it too. Especially all I wanted was an image to run crunch on, small (no need more) but with enough details to check for the output. I used a 64px unvanquished icon instead: it is small but detailed enough, and features alpha channel with some shadows, so it looks complex enough to me. It is not displayed, just stored in `test/`.

**Comment by illwieckz** [comment]
So, now Linux amd64, macOS amd64 and Winsows/MSVC run the tests. The tests are only about running crunch on a set of images and only to check that crunch doesn't error out. We may do some checksums later but that's good enough for now and it's enough to validate the infrastructure is there.

**Comment by illwieckz** [comment]
For some unknown reason, when running in azure pipeline, `test/test.py` own `print` is not displayed in azure pipeline output, same problem with `stdout` and `stderr`, I have no idea why. That's not a big problem and this may be investigated separately.

**Pull request merged**